File: packages/shared/src/lab-queue.ts
/** Lab queue helpers — parse comma-separated orders and classify HC III tests. */

import { LAB_CATALOG_TESTS } from './lab-catalog';

export type LabTestStatus = 'pending' | 'running' | 'done' | 'abnormal';

export interface LabTestResultRow {
  test: string;
  status: LabTestStatus;
  result: string | null;
  abnormal: boolean;
  started_at?: string | null;
  completed_at?: string | null;
}

export function parseTestsOrdered(testsOrdered: string | null | undefined): string[] {
  if (!testsOrdered?.trim()) return [];
  return testsOrdered
    .split(',')
    .map((s) => s.trim())
    .filter(Boolean);
}

/**
 * Union of an existing comma-separated tests_ordered value with newly ordered
 * tests. Existing order is preserved, incoming tests are appended, duplicates
 * are dropped case-insensitively (first occurrence's casing wins). Returns
 * null when both inputs are empty — the "no order" representation used by the
 * lab board.
 *
 * SQL mirror: merge_tests_ordered() in
 * packages/supabase/migrations/108_lab_order_protection_and_queue_release.sql —
 * keep the two in sync. Lab orders are ADD-ONLY through the note/order paths
 * (LAB-1 regression: the note editor's stale snapshot must never erase an
 * order submitted from another surface).
 */
export function mergeTestsOrdered(
  existing: string | null | undefined,
  incoming: string | string[] | null | undefined,
): string | null {
  const incomingList = Array.isArray(incoming)
    ? incoming.map((s) => s.trim()).filter(Boolean)
    : parseTestsOrdered(incoming);
  const seen = new Set<string>();
  const out: string[] = [];
  for (const name of [...parseTestsOrdered(existing), ...incomingList]) {
    const key = name.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(name);
  }
  return out.length > 0 ? out.join(', ') : null;
}

/** Visit-level lab statuses that keep a visit on the lab bench. */
export const LAB_QUEUE_OPEN_STATUSES = ['pending', 'running'] as const;

/**
 * Lab queue membership: the visit-level status is open AND at least one
 * ordered test has not reached a terminal state. Mirrors the SQL filter in
 * apps/web/src/app/dashboard/lab/page.tsx (lab_status IN pending/running)
 * plus its countOpenLabTests post-filter.
 */
export function isLabQueueVisit(visit: {
  tests_ordered: string | null | undefined;
  lab_status: string | null | undefined;
  lab_test_results?: LabTestResultRow[] | null;
}): boolean {
  if (
    !visit.lab_status ||
    !(LAB_QUEUE_OPEN_STATUSES as readonly string[]).includes(visit.lab_status)
  ) {
    return false;
  }
  const tests = mergeLabTestResults(visit.tests_ordered, visit.lab_test_results ?? []);
  return countOpenLabTests(tests) > 0;
}

export function mergeLabTestResults(
  testsOrdered: string | null | undefined,
  stored: LabTestResultRow[] | null | undefined,
): LabTestResultRow[] {
  const names = parseTestsOrdered(testsOrdered);
  const byName = new Map((stored ?? []).map((r) => [r.test, r]));
  return names.map((test) => {
    const existing = byName.get(test);
    if (existing) return existing;
    return { test, status: 'pending', result: null, abnormal: false };
  });
}

/** Qualitative bench tests get one-tap Positive / Negative chips at HC III. */
export function labTestSupportsPosNeg(testName: string): boolean {
  const normalized = testName.toLowerCase();
  const catalog = LAB_CATALOG_TESTS.find(
    (t) => t.name.toLowerCase() === normalized || t.code.toLowerCase() === normalized,
  );
  if (catalog) {
    const qualitative = new Set([
      'MRDT',
      'BS_MPS',
      'HIV_RDT',
      'SYPHILIS',
      'UCG',
      'WIDAL',
      'STOOL_RDT',
      'AFB',
    ]);
    return qualitative.has(catalog.code);
  }
  return /malaria|hiv|pregnancy|ucg|syphilis|rpr|tpha|widal|typhoid|h\.?\s*pylori|rdt|rapid|afb|tb\s*smear|sputum/i.test(
    testName,
  );
}

export function countOpenLabTests(rows: LabTestResultRow[]): number {
  return rows.filter((r) => r.status === 'pending' || r.status === 'running').length;
}

export function formatLabResultsSummary(rows: LabTestResultRow[]): string {
  return rows
    .filter((r) => r.status === 'done' || r.status === 'abnormal')
    .map((r) => `${r.test}: ${r.result ?? '—'}`)
    .join('; ');
}

File: apps/android/app/src/main/java/com/karibuhealth/app/domain/LabQueue.kt
package com.karibuhealth.app.domain

import com.karibuhealth.app.domain.model.LabTestResultRow
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** Per-test lab queue helpers — mirrors packages/shared/src/lab-queue.ts */
object LabQueue {

    private val json = Json { ignoreUnknownKeys = true }

    data class DerivedLabState(
        val labStatus: String,
        val labResults: String?,
        val labAbnormal: Boolean,
        val allComplete: Boolean,
    )

    fun parseTestsOrdered(testsOrdered: String?): List<String> {
        if (testsOrdered.isNullOrBlank()) return emptyList()
        return testsOrdered.split(",").map { it.trim() }.filter { it.isNotEmpty() }
    }

    fun parseStoredResults(jsonStr: String?): List<LabTestResultRow> {
        if (jsonStr.isNullOrBlank() || jsonStr == "[]") return emptyList()
        return runCatching {
            json.decodeFromString<List<StoredRow>>(jsonStr).map { it.toRow() }
        }.getOrDefault(emptyList())
    }

    fun encodeResults(rows: List<LabTestResultRow>): String =
        json.encodeToString(rows.map { StoredRow.from(it) })

    fun mergeLabTestResults(
        testsOrdered: String?,
        stored: List<LabTestResultRow>,
    ): List<LabTestResultRow> {
        val byName = stored.associateBy { it.test }
        return parseTestsOrdered(testsOrdered).map { name ->
            byName[name] ?: LabTestResultRow(test = name, status = "pending")
        }
    }

    fun countOpenTests(rows: List<LabTestResultRow>): Int =
        rows.count { it.status == "pending" || it.status == "running" }

    fun labTestSupportsPosNeg(testName: String): Boolean {
        val normalized = testName.lowercase()
        val catalog = LabCatalog.tests.find {
            it.name.lowercase() == normalized || it.code.lowercase() == normalized
        }
        if (catalog != null) {
            return catalog.code in QUALITATIVE_CODES
        }
        return Regex(
            "malaria|hiv|pregnancy|ucg|syphilis|rpr|tpha|widal|typhoid|h\\.?\\s*pylori|rdt|rapid|afb|tb\\s*smear|sputum",
            RegexOption.IGNORE_CASE,
        ).containsMatchIn(testName)
    }

    fun applyStartTest(rows: List<LabTestResultRow>, testName: String): List<LabTestResultRow> =
        rows.map { row ->
            if (row.test == testName && row.status == "pending") {
                row.copy(status = "running")
            } else {
                row
            }
        }

    fun applyRecordResult(
        rows: List<LabTestResultRow>,
        testName: String,
        result: String,
        abnormal: Boolean,
    ): List<LabTestResultRow> =
        rows.map { row ->
            if (row.test == testName) {
                row.copy(
                    status = if (abnormal) "abnormal" else "done",
                    result = result,
                    abnormal = abnormal,
                )
            } else {
                row
            }
        }

    fun deriveVisitLabState(rows: List<LabTestResultRow>): DerivedLabState {
        if (rows.isEmpty()) {
            return DerivedLabState("not_ordered", null, false, true)
        }
        val done = rows.count { it.status == "done" || it.status == "abnormal" }
        val abnormalCount = rows.count { it.abnormal || it.status == "abnormal" }
        val anyRunning = rows.any { it.status == "running" }
        val summary = rows
            .filter { it.status == "done" || it.status == "abnormal" }
            .joinToString("; ") { "${it.test}: ${it.result ?: "—"}" }
            .ifBlank { null }

        val status = when {
            done == 0 && !anyRunning -> "pending"
            done < rows.size -> "running"
            abnormalCount > 0 -> "abnormal"
            else -> "done"
        }
        return DerivedLabState(
            labStatus = status,
            labResults = summary,
            labAbnormal = abnormalCount > 0,
            allComplete = done == rows.size,
        )
    }

    private val QUALITATIVE_CODES = setOf(
        "MRDT", "BS_MPS", "HIV_RDT", "SYPHILIS", "UCG", "WIDAL", "STOOL_RDT", "AFB",
    )

    @Serializable
    private data class StoredRow(
        val test: String,
        val status: String,
        val result: String? = null,
        val abnormal: Boolean = false,
    ) {
        fun toRow() = LabTestResultRow(test, status, result, abnormal)
        companion object {
            fun from(row: LabTestResultRow) = StoredRow(row.test, row.status, row.result, row.abnormal)
        }
    }
}
